### PR TITLE
Create one regular user for running tests

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
Will allow https://github.com/sociomantic-tsunami/turtle/pull/19 to switch back to regular image.

### CODE REVIEW

**Comment by leandro-lucarella-sociomantic** [comment]
This is a bit tricky because it's very dependant on the environment used.

I guess this comes from: https://github.com/sociomantic-tsunami/turtle/pull/19/files#diff-1d37e48f9ceff6d8030570cd36286a61R18

What I don't understand is how CircleCI shares the code with the docker instance, because if it's via a volume, then the user ID needs to match in the host and in the container. This was a problem I always had with creating users in docker images, that you need some instrumentation to make the UIDs match with the host system.

BIkeshed: I will name the user either "cachalot" or maybe "sudo" to make clear that the user should be probably only used for sudo access.

**Comment by mihails-strasuns-sociomantic** [comment]
> What I don't understand is how CircleCI shares the code with the docker instance

Checkout happens within the container context so there is nothing to share. https://circleci.com/docs/2.0/configuration-reference/#checkout says that they do some tricks like detecting if `git` is present in the container image so it is not a completely honest `git clone`, but principle is the same.

> need some instrumentation to make the UIDs match with the host system

Only case I can imagine when you may need that is if you produce build artifacts in mounted folder, don't use docker volumes and permissions/ownership of those files truly matter. Which is both very unidiomatic for docker and unapplicable to CircleCI case.

Am I missing something?

**Comment by mihails-strasuns-sociomantic** [comment]
> BIkeshed: I will name the user either "cachalot" or maybe "sudo" to make clear that the user should be probably only used for sudo access.

As you wish :)

**Comment by leandro-lucarella-sociomantic** [comment]
> Checkout happens within the container context so there is nothing to share. https://circleci.com/docs/2.0/configuration-reference/#checkout says that they do some tricks like detecting if git is present in the container image so it is not a completely honest git clone, but principle is the same.

Ah, interesting, so they rely on the image having git? On the other hand is weird, I guess some people wouldn't want to install git on an image if it's used for deployment for example.

> Am I missing something?

I guess not.

But please also include some mention to the new user in the README. I would go with `cachalot` user + explaining in the README that this user has sudo access.

**Comment by mihails-strasuns-sociomantic** [comment]
> Ah, interesting, so they rely on the image having git?

There was a mention that steps falls back to fetching their own git binary if not present in the image but I can't find it now :(

**Comment by mihails-strasuns-sociomantic** [comment]
Updated

**Comment by leandro-lucarella-sociomantic** [comment]
Ah, one more request. Add a test :innocent:

**Comment by mihails-strasuns-sociomantic** [comment]
Added one very simple one (`id cachalot`) - not sure if there is anything else to test.

**Comment by llucax** [comment]
> Added one very simple one (`id cachalot`) - not sure if there is anything else to test.

Test that sudo is working for that user?

	docker run -u cachalot sudo -n true

(untested, but you get the idea)

**Comment by mihails-strasuns-sociomantic** [comment]
Actually even `id cachalot` doesn't seem to work, looking at CI failure. Maybe I am misunderstanding something about the test system?

**Comment by llucax** [comment]
That's weird, the test does look good to me (the code).

**Comment by mihails-strasuns-sociomantic** [comment]
Weird thing is that if I run exactly same command for generated image on my machine, `id cachalot` works as expected.

**Comment by nemanja-boric-sociomantic** [comment]
Ah, I see - the tests are executed with the tagged image, not the `latest`:

```
+docker run -t --rm -e HOME=/tmp -w /tmp -v /home/travis/build/sociomantic-tsunami/cachalot/test:/tmp/test sociomantictsunami/base:xenial-v3 test/base-root

+ id cachalot

id: 'cachalot': no such user
```

It's using tagged `xenial-v3` instead of `latest`.

**Comment by nemanja-boric-sociomantic** [comment]
Hmm, but you still tag it:

```
Successfully built 6fda2d781226
docker tag "sociomantictsunami/base:xenial-v3" "sociomantictsunami/base"
```

However, look into the docker images output (bottom of the file):

```
$ docker images

sociomantictsunami/dlang   latest              e4b022b43219        28 seconds ago       941 MB

sociomantictsunami/dlang   xenial-v3           e4b022b43219        28 seconds ago       941 MB

sociomantictsunami/base    latest              6fda2d781226        About a minute ago   583 MB

ubuntu                     xenial              20c44cd7596f        8 days ago           123 MB

sociomantictsunami/base    xenial-v3           66df9637c923        2 months ago         580 MB
```

So, something with the tagging doesn't work.

**Comment by nemanja-boric-sociomantic** [comment]
Oh, is the `docker tag` command other way around? Meaning, should it be:

```
docker tag "sociomantictsunami/base" "sociomantictsunami/base:xenial-v3"
```
instead 

```
docker tag "sociomantictsunami/base:xenial-v3" "sociomantictsunami/base"
```

```
nemanjaboric@labs-129:/home/nemanjaboric/work/propulsor  git:(pindmd*) $ docker help tag

Usage:	docker tag SOURCE_IMAGE[:TAG] TARGET_IMAGE[:TAG]

Create a tag TARGET_IMAGE that refers to SOURCE_IMAGE

```

**Comment by leandro-lucarella-sociomantic** [comment]
> Oh, is the docker tag command other way around?

I don't think so, the image is created as `sociomantictsunami/base:xenial-v3`:
```sh
docker build --pull --build-arg VERSION_IMAGE=v3 -t sociomantictsunami/base:xenial-v3 ...
```
(see the -t)

I think this is getting really confusing because of the `-j2`. I suggest removing the `-j2` from `make` in the `.travis.yml` temporarily to debug this problem (or trying it out locally :).

**Comment by Burgos** [comment]
Right, but  `docker images` doesn't show this freshly built image as the tagged one (instead it's `latest`).

And see what's going on when building the dlang image - it thinks the image on the hub is newer:

```
 ---> 6fda2d781226 # This is the new xenial-v3 image that we just built

Removing intermediate container d1b7089fc138
Successfully built 6fda2d781226
docker tag "sociomantictsunami/base:xenial-v3" "sociomantictsunami/base"
==================================================
./build-img  "sociomantictsunami/dlang:xenial-v3"
+docker build --pull --build-arg VERSION_IMAGE=v3 -t sociomantictsunami/dlang:xenial-v3 -f dlang.Dockerfile.v3-xenial.tmp .
Sending build context to Docker daemon 20.48 kB
Step 1/6 : FROM sociomantictsunami/base:xenial-v3
xenial-v3: Pulling from sociomantictsunami/base
Status: Downloaded newer image for sociomantictsunami/base:xenial-v3 # wtf?
 ---> 66df9637c923
````

**Comment by Burgos** [comment]
> I think this is getting really confusing because of the -j2. I suggest removing the -j2 from make in the .travis.yml temporarily to debug this problem (or trying it out locally :).

🤦‍♂️

**Comment by leandro-lucarella-sociomantic** [comment]
I'm running some test locally now :)

**Comment by leandro-lucarella-sociomantic** [comment]
OK, the problem seems to be definitely `--pull`. From one side it is broken because if the tag doesn't already exist in docker hub, it fails, which means build for new branches will always fail.

I guess then it must be prioritizing the pulled image even if older, which actually smells like a docker bug to me.

For example: https://travis-ci.org/leandro-lucarella-sociomantic/cachalot/builds/308600566#L1828

**Comment by leandro-lucarella-sociomantic** [comment]
Rebase on top of https://github.com/sociomantic-tsunami/cachalot/pull/40 to see if this fixes the problem please.

**Comment by leandro-lucarella-sociomantic** [comment]
It seems to work: https://travis-ci.org/leandro-lucarella-sociomantic/cachalot/builds/308649826

**Comment by mihails-strasuns-sociomantic** [comment]
Yeah works after rebase. I have also added tests for internal user.

**Comment by leandro-lucarella-sociomantic** [comment]
It fails with permissions denied, I suggest to debug locally.

**Comment by mihails-strasuns-sociomantic** [comment]
Forgot to chmod, green now.

**Review by leandro-lucarella-sociomantic — COMMENTED** [review]
Add a `relnote` and is good to go from my side.

**Comment by mihails-strasuns-sociomantic** [comment]
Updated

**Comment by leandro-lucarella-sociomantic** [comment]
I can't see a release note... :thinking:

**Comment by mihails-strasuns-sociomantic** [comment]
Pushed to wrong branch, here it is now :)

File: README.rst [hunk 1]
```diff
@@ -23,3 +23,13 @@ a `-`), the default tagged image will be `ORG/IMG:DIST-vX-test` (instead of
...
+
+Internal `cachalot` user
+========================
+
+As part of base image generation, user named `cachalot` will be created. It is
+intended to be used to run tests without accidental root access which may affect
+the result.
+
+This user is added to sudoers with a permission to run any sudo command without
+a password.
```
Comment (leandro-lucarella-sociomantic) [inline README.rst]: Sorry, I don't know how I didn't catch this earlier, but I think the main intention of this user is to not have to build a new image just to install dependencies, and being able to use `sudo apt install` as part of the build process instead.  Normally test were ran as users anyway. Maybe that should be mentioned in the readme and release notes.

Comment (mihails-strasuns-sociomantic) [reply README.rst]: Updated. Like this?

**Review by leandro-lucarella-sociomantic — APPROVED** [review]

**Pull request merged**